src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,371 @@
+            }
+            foreach (var providerReference in this._namedProviderReferences.Values)
+            {
+                if (providerReference == provider)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private bool IsProviderActive(FeatureProviderReference providerRef)
+        {
+            return this._activeSubscriptions.Contains(providerRef);
+        }
+
+        private void EmitOnRegistration(FeatureProviderReference provider, ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            if (provider == null)
+            {
+                return;
+            }
+            var status = provider.Provider.GetStatus();
+
+            var message = "";
+            if (status == ProviderStatus.Ready && eventType == ProviderEventTypes.ProviderReady)
+            {
+                message = "Provider is ready";
+            }
+            else if (status == ProviderStatus.Error && eventType == ProviderEventTypes.ProviderError)
+            {
+                message = "Provider is in error state";
+            }
+            else if (status == ProviderStatus.Stale && eventType == ProviderEventTypes.ProviderStale)
+            {
+                message = "Provider is in stale state";
+            }
+
+            if (message != "")
+            {
+                try
+                {
+                    handler.Invoke(new ProviderEventPayload
+                    {
+                        ProviderName = provider.Provider?.GetMetadata()?.Name,
+                        Type = eventType,
+                        Message = message
+                    });
+                }
+                catch (Exception exc)
+                {
+                    this.Logger?.LogError("Error running handler: " + exc);
+                }
+            }
+        }
+
+        private async void ProcessFeatureProviderEventsAsync(object providerRef)
+        {
+            while (true)
+            {
+                var typedProviderRef = (FeatureProviderReference)providerRef;
+                if (typedProviderRef.Provider.GetEventChannel() == null)
+                {
+                    return;
+                }
+                var item = await typedProviderRef.Provider.GetEventChannel().Reader.ReadAsync().ConfigureAwait(false);
+
+                switch (item)
+                {
+                    case ProviderEventPayload eventPayload:
+                        await this.EventChannel.Writer.WriteAsync(new Event { Provider = typedProviderRef, EventPayload = eventPayload }).ConfigureAwait(false);
+                        break;
+                    case ShutdownSignal _:
+                        typedProviderRef.ShutdownSemaphore.Release();
+                        return;
+                }
+            }
+        }
+
+        // Method to process events
+        private async void ProcessEventAsync()
```

Comment:
Given we execute this in a thread, do we need this method to be async anymore?

Reply:
also check asyn requirement on `private async void ProcessFeatureProviderEventsAsync(object providerRef)` method

Reply:
I think this still needs to be `async` since there's an `await` being used within the method - removing the `async` will cause the compiler to complain

Reply:
Thanks , will resolve this conversation